firedrake/slate/slate.py

```diff
@@ -0,0 +1,767 @@
+        self._domain_numering = dict((d, i) for i, d in enumerate(self._integral_domains))
+
+        self._hash = None
+        self._signature = None
+
+    # Overloaded operations for the tensor algebra
+    def __add__(self, other):
+        return TensorAdd(self, other)
+
+    def __radd__(self, other):
+        """Ordering of tensor addition does not matter."""
+        return self.__add__(other)
+
+    def __sub__(self, other):
+        return TensorSub(self, other)
+
+    def __rsub__(self, other):
+        """Ordering of tensor subtraction does not matter."""
+        return self.__sub__(other)
+
+    def __mul__(self, other):
+        return TensorMul(self, other)
+
+    def __rmul__(self, other):
+        """Tensor multiplication is not commutative in general."""
+        return TensorMul(other, self)
+
+    def __neg__(self):
+        return Negative(self)
+
+    def __pos__(self):
+        return Positive(self)
+
+    def equals(self, other):
+        """Evaluates the boolean expression `A == B`."""
+
+        if not isinstance(other, Tensor):
+            return False
+        if len(self.arguments()) != len(other.arguments()):
+            return False
+        if len(self.coefficients()) != len(other.coefficients()):
+            return False
+        if len(self._integrals) != len(other._integrals):
+            return False
+        return (self._arguments == other._arguments and
+                self._coefficients == other._coefficients and
+                self._integrals == other._integrals)
+
+    def __ne__(self, other):
+        return not self.equals(other)
+
+    def __eq__(self, other):
+        """Evaluation of the "==" operator using the SLATE class: SlateEquation."""
+        return SlateEquation(self, other)
+
+    # Essential properties
+    @property
+    def inv(self):
+        return Inverse(self)
+
+    @property
+    def T(self):
+        return Transpose(self)
+
+    @property
+    def operands(self):
+        """Returns the tensor objects which this tensor operates on."""
+        return ()
+
+    # Analyze integrals of tensor
+    @classmethod
+    def check_integrals(cls, integrals):
+        """Checks for positive restrictions on integrals."""
+
+        mapper = CheckRestrictions()
+        for it in integrals:
+            map_integrand_dags(mapper, it)
+
+    # Accessor methods
+    def arguments(self):
+        """Returns a tuple of arguments associated with the tensor."""
+        return self._arguments
+
+    def coefficients(self):
+        """Returns a tuple of coefficients associated with the tensor."""
+        return self._coefficients
+
+    def get_ufl_integrals(self):
+        """Returns the associated ufl integrals with the tensor
+        (This is necessary for form compilation).
+
+        This function should really only be called on base-Tensor objects
+        like an individual Vector or Matrix rather than a full SLATE expr."""
+        return self._integrals
```

Comment:
To ensure your comment is obeyed, you can `raise ValueError` on the subclasses for which this doesn't make sense.